### DISCUSSION TITLE
Guidance on Implementing NKey Authentication with the Swift NATS SDK

### DISCUSSION THREAD

**Discussion by dmaulikr** [nats-io/nats.swift#63] [Q&A] [answered]
Hello,

We're currently transitioning from a workaround where we wrapped the C library for NATS in Swift to directly using the nats.io Swift SDK for our project. Previously, we used a combination of a public key, a server URL, and a seed file (seed.nk) for NKey authentication and connection. This method is documented here: https://github.com/dmaulikr/NatsCSwift.

As we're now adopting the Swift SDK, we're looking for advice on properly implementing NKey authentication using the mentioned assets. Specifically, we're interested in connecting to the server and sending messages with our existing credentials. Any examples or guidance on this would be greatly appreciated, as we aim to leverage the Swift SDK's capabilities fully without resorting to our previous workaround, which has led to occasional issues.

Thank you for your assistance.

Thank you!

**Comment by Jarema** [comment]
Hello!

You can use crede  `credentialsFile` option:

```
    func testCredentialsAuth() async throws {
        logger.logLevel = .debug
        let bundle = Bundle.module
        natsServer.start(cfg: bundle.url(forResource: "jwt", withExtension: "conf")!.relativePath)

        let creds = bundle.url(forResource: "TestUser", withExtension: "creds")!

        let client = NatsClientOptions().url(URL(string: natsServer.clientURL)!).credentialsFile(
            creds
        ).build()
        try await client.connect()
        let subscribe = try await client.subscribe(subject: "foo").makeAsyncIterator()
        try await client.publish("data".data(using: .utf8)!, subject: "foo")
        _ = await subscribe.next()
    }
```

which uses [.creds file](https://github.com/nats-io/nats.swift/blob/main/Tests/NatsTests/Integration/Resources/TestUser.creds) as a standard pattern, when generating creds from `nsc` tool. Creds file includes JWT and seed.

We can include option to provide those values separately.

**Reply by dmaulikr** [reply]
Hello @Jarema,

Thank you very much for your prompt response. We're currently facing a challenge as we do not have a JWT provided by our end client. Our available resources are only the URL, Public Key, and Seed. Attempting to connect to the NATS server without a JWT, as expected, results in failure. Our previous solution involved using the NATS C library with `natsOptions_SetNKeyFromSeed`, which allowed for connection using just a public key and seed file.

Could you advise if there's a way to achieve a similar connection using only a public key and seed with the current Swift SDK?

We appreciate your guidance on this matter.

Thank you!

**Reply by Jarema** [reply]
you are right. We will add this method in next release.

**Reply by Jarema** [reply]
The PR https://github.com/nats-io/nats.swift/pull/65 has     been merged and will be part of the next release.
Note that you do not need public key if you pass seed. Public key can be derived from that seed key.

**Reply by dmaulikr** [reply]
Hello @Jarema,

Thank you for your support. After implementing the updates from the main branch of the Swift package, I am encountering a crash within the library itself. The logs show the following error:
**Nats/NatsConnection.swift:404: Fatal error: leaking promise created at (file: "Nats/NatsConnection.swift", line: 404)**
The crash occurs in **swift-nio -> sources -> NIOCore -> EventLoopFuture, specifically at line 425: fatalError("leaking promise created at \(creation)", file: creation.file, line: creation.line).**

I've tested the connection using two different methods. Both methods result in the same crash at the location mentioned above. Here are the methods for reference:

Common method:
```
let nats = NatsClientOptions().url(URL(string: serverURL)!).build()
Task.init {
    do {
        try await nats.connect()
    } catch {
        print("Failed to connect to NATS server: \(error)")
    }
}
```
Main method using NKey:
```
let client = NatsClientOptions().url(URL(string: serverURL)!).nkey(strSeedKey).build()
Task.init {
    do {
        try await client.connect()
    } catch {
        print("Failed to connect to NATS server: \(error)")
    }
}
```
Could you please provide guidance on how to address this issue? or there's anything that I am missing?

Thank you!

**Reply by Jarema** [reply]
We will take a look into this.

**Reply by mtmk** [reply, marked as answer]
@dmaulikr we have a fix in #66 are you able to try that?
edit: just merged. it should be in main now

**Comment by dmaulikr** [comment]
@Jarema @mtmk, It is working now. Thank you for looking into this request.

Thank you for support.
